lib/vrl/stdlib/src/encode_logfmt.rs

```diff
@@ -0,0 +1,307 @@
+use std::collections::BTreeMap;
+use std::fmt::Write;
+use std::result::Result;
+use vrl::prelude::*;
+use Value::{Array, Object};
+
+#[derive(Clone, Copy, Debug)]
+pub struct EncodeLogfmt;
+
+impl Function for EncodeLogfmt {
+    fn identifier(&self) -> &'static str {
+        "encode_logfmt"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::OBJECT,
+                required: true,
+            },
+            Parameter {
+                keyword: "fields_ordering",
+                kind: kind::ARRAY,
+                required: false,
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+        let fields = arguments.optional("fields_ordering");
+
+        Ok(Box::new(EncodeLogfmtFn { value, fields }))
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "encode object",
+                source: r#"encode_logfmt!({"lvl": "info", "msg": "This is a message", "log_id": 12345})"#,
+                result: Ok(r#"s'log_id=12345 lvl=info msg="This is a message"'"#),
+            },
+            Example {
+                title: "encode object with fields ordering",
+                source: r#"encode_logfmt!({"msg": "This is a message", "lvl": "info", "log_id": 12345}, ["lvl", "msg"])"#,
+                result: Ok(r#"s'lvl=info msg="This is a message" log_id=12345'"#),
+            },
+        ]
+    }
+}
+
+#[derive(Clone, Debug)]
+struct EncodeLogfmtFn {
+    value: Box<dyn Expression>,
+    fields: Option<Box<dyn Expression>>,
+}
+
+fn resolve_fields(fields: Value) -> Result<Vec<String>, ExpressionError> {
+    let arr = fields.try_array()?;
+    arr.iter()
+        .enumerate()
+        .map(|(idx, v)| {
+            v.try_bytes_utf8_lossy()
+                .map(|v| v.to_string())
+                .map_err(|e| format!("invalid field value type at index {}: {}", idx, e).into())
+        })
+        .collect()
+}
+
+impl Expression for EncodeLogfmtFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let fields = match &self.fields {
+            None => Ok(vec![]),
+            Some(expr) => {
+                let fields = expr.resolve(ctx)?;
+                resolve_fields(fields)
+            }
+        }?;
+
+        let object = value.try_object()?;
+        Ok(encode(object, &fields[..]).into())
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value.type_def(state).fallible().bytes()
```

Comment:
nit:

The output of this function doesn't depend on its input (which you can see since we're setting both the faillibity (`fallible()`) and the type (`bytes()`) manually), so to avoid any confusion, I'd write this as:

```rust
TypeDef::bytes().fallible()
```

Reply:
6a22ec0